### DISCUSSION TITLE
SupportsRichComparison issue in latest version of Pylance

### DISCUSSION THREAD

**Discussion by ColorfulQuark** [microsoft/pyright#5660] [Q&A]
Code that was not flagged in prior versions of Pylance now gets:

Expression of type "tuple[dict[str, Quote], dict[str, SupportsRichComparison | date]]" cannot be assigned to return type "tuple[dict[str, Quote], dict[str, date | datetime]]"

How should I deal with this?

**Comment by erictraut** [comment]
Could you provide a minimal, self-contained code sample that demonstrates your problem? I'm unable to diagnose your problem or suggest a solution from your description above.

**Reply by ColorfulQuark** [reply]
`from datetime import date, datetime

def early_late_x():
    dts = [date(1900, 1, 1), datetime(1900, 1, 1)]
    return min(dts), max(dts)


def get_quotes_x() -> dict[str, date | datetime]:
    earliest, latest = early_late_x()

    middle = date(1990, 1, 1)

    return {
        "earliest": earliest,
        "latest": latest,
        "middle": middle,
    }
`
`Expression of type "dict[str, SupportsRichComparison | date]" cannot be assigned to return type "dict[str, date | datetime]"` This points to the return statement.

**Reply by erictraut** [reply]
By default, [pyright's inference rules for list expressions](https://microsoft.github.io/pyright/#/type-inference?id=list-expressions) produce a `list[Unknown]` if the list expression contains heterogeneous types like in your example. That means `dts` has an inferred type of `list[Unknown]`, and `early_late_x` has an inferred return type of `tuple[SupportsRichComparison, SupportsRichComparison]`. 

You can fix this by doing one of the following:
1. Specify that you would like to use "strictListInference", which changes the inference rules for list expressions. Pylance doesn't expose this as a configuration option, but you can specify it using a [pyrightconfig.json or pyproject.toml file](https://microsoft.github.io/pyright/#/configuration?id=main-configuration-options). You can also specify it on a per-file basis using a `# pyright: strictListInference=true` comment at the top of the file.
2. You can provide an explicit type annotation for `dts`.
```python
dts: list[date | datetime] = ...
```
3. If you find that you aren't getting sufficient value from static type checking, you can disable it entirely by setting `python.analsysi.typeCheckingMode` to "off".

My recommended approach is option 2 above.

**Reply by ColorfulQuark** [reply]
That worked thanks.

Out of curiosity, did it change in the latest version?  I wasn't getting this issue before.
